src/providers/aws.rs

```diff
@@ -1302,22 +1302,39 @@ impl RegionLauncher {
             .await;
         }
 
+        use rusoto_core::RusotoError;
         if !self.security_group_id.trim().is_empty() {
-            let group_span = tracing::trace_span!("security_group", id = %self.security_group_id);
+            let group_span =
+                tracing::trace_span!("removing security group", id = %self.security_group_id);
             async {
                 tracing::trace!("removing security group.");
                 // clean up security groups and keys
+                let start = tokio::time::Instant::now();
                 loop {
+                    if start.elapsed() > tokio::time::Duration::from_secs(60) {
+                        tracing::warn!(
+                            "failed to clean up temporary security group after 60 seconds."
+                        );
+                    }
+
                     let mut req = rusoto_ec2::DeleteSecurityGroupRequest::default();
                     req.group_id = Some(self.security_group_id.clone());
-                    if let Err(e) = client.delete_security_group(req).in_current_span().await {
-                        tracing::trace!(
-                            "failed to clean up temporary security group: {}. retrying.",
-                            e,
-                        );
-                        tokio::time::delay_for(tokio::time::Duration::from_secs(5)).await;
-                    } else {
-                        break;
+                    match client.delete_security_group(req).await {
+                        Ok(_) => break,
+                        Err(RusotoError::Unknown(r)) => {
+                            let err = r.body_as_str();
+                            if err.contains("<Code>DependencyViolation</Code>") {
+                                tracing::trace!("instances not yet shut down -- retrying");
+                                tokio::time::delay_for(tokio::time::Duration::from_secs(5)).await;
+                            } else {
+                                tracing::warn!("failed: {}", err);
```

Comment:
Here I think we actually want to return an error.

Reply:
Also the timeout case above this needs to be addressed. 
So actually this method currently doesn't return `Result<(), Error>`, it returns `()`. Should we make it return `Result`, in case of timeout? I think this error is not actually fatal (since obviously we have been dealing with failed deletions all this time), and might be fine to keep as a warning (vs users who might be using `?`)

Reply:
Hmm, it's a good question... I _feel_ like the right thing to do is to have `terminate_all` also return a `Result`. The user is free to ignore errors from `terminate_all` if they wish, but it seems to me like the right thing to do is to at least _report_ that there was an error. That way we can also start to properly return errors for, say, when terminating the instances fails!

Reply:
Alright, I have written a version that collects all the errors without short-circuiting on failures (so that everything that can be cleaned up, is cleaned up). I've probably used this new `Report`/`eyre` stuff incorrectly (tried to model from the baremetal collecting errors example)...